cmds/tail/tail.go

```diff
@@ -0,0 +1,261 @@
+	_, err = writer.Write(data)
+	if err != nil {
+		return err
+	}
+	// reposition the stream at the end of what we've read
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
+	return nil
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
+	if err != nil {
+		return err
+	}
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
+	var slice []byte             // will hold the final data, after moving line by line
+	for {
+		n, err := io.ReadFull(&inFile, buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
+				return err
+			}
+		}
+		// look for newlines and keep a slice starting at the n-th to last line
+		// (no further than numLines)
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
+		foundNewLines := bytes.Count(slice, []byte{'\n'})
+		newLinesToSkip := 0
+		if foundNewLines > numLines {
+			newLinesToSkip = foundNewLines - numLines
+		}
+		// if there are more than `numLines` lines, skip them
+		for idx := 0; idx < newLinesToSkip; idx++ {
+			pos := bytes.IndexByte(slice, '\n')
+			// in theory the code below should never panic for out of bounds because
+			// of the checks above, but..
+			if len(slice) < pos+1 {
+				return fmt.Errorf("less bytes than expected were found")
+			}
+			slice = slice[pos+1:]
+		}
+	}
+	_, err = writer.Write(slice)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func Tail(inFile os.File, writer io.Writer, config TailConfig) error {
+	var (
+		err                error
+		retryFromBeginning = false
+	)
+	if config.numLines < 0 {
+		return fmt.Errorf("The number of lines cannot be negative")
+	}
+	// try reading from the end of the file
+	err = readLastLinesBackwards(inFile, writer, config.numLines)
+	if err != nil {
+		// if it failed because it couldn't seek, mark it for retry reading from
+		// the beginning
+		if pathErr, ok := err.(*os.PathError); ok && pathErr.Err.Error() == "illegal seek" {
+			retryFromBeginning = true
+		} else {
+			return err
+		}
+	}
+	// if reading backwards failed because the file is not seekable,
+	// retry from the beginning
+	if retryFromBeginning {
+		err = readLastLinesFromBeginning(inFile, writer, config.numLines)
+	}
+	if config.follow {
```

Comment:
Do you want to perform the tail and then throw an error about `-f` not being implemented or throw an error up front? As is this will perform the tail then throw an error. I'm not sure which is more pleasant for someone using `tail`.

Reply:
yeah, it's a better user experience to bail out immediately, I've fixed this